src/apify/_utils.py

```diff
@@ -410,3 +412,118 @@ def validate_single(field_value: Any, expected_type: type, required: bool, name:
 PARSE_DATE_FIELDS_MAX_DEPTH = 3
 PARSE_DATE_FIELDS_KEY_SUFFIX = 'At'
 ListOrDictOrAny = TypeVar('ListOrDictOrAny', list, dict, Any)
+
+
+def get_short_base64_hash(data: bytes) -> str:
+    """Generates an 8-character, base64-encoded SHA-256 hash of the provided data.
+
+    Args:
+        data: The binary data to be hashed.
+
+    Returns:
+        An 8-character, base64-encoded hash.
+    """
+    # Create a SHA-256 hash of the payload.
+    hash_object = sha256(data)
+    # Convert the hash to a base64-encoded string.
+    base64_encoded_hash = b64encode(hash_object.digest()).decode()
+    # Remove any '+' or '/' or '=' characters and return the first 8 characters.
+    return base64_encoded_hash.replace('+', '').replace('/', '').replace('=', '')[:8]
+
+
+def normalize_url(url: str, *, keep_url_fragment: bool = False) -> str:
+    """Normalizes a URL.
+
+    This function cleans and standardizes a URL by removing leading and trailing whitespaces,
+    converting the scheme and netloc to lower case, stripping unwanted tracking parameters
+    (specifically those beginning with 'utm_'), sorting the remaining query parameters alphabetically,
+    and optionally retaining the URL fragment. The goal is to ensure that URLs that are functionally
+    identical but differ in trivial ways (such as parameter order or casing) are treated as the same.
+
+    Args:
+        url: The URL to be normalized.
+        keep_url_fragment: Flag to determine whether the fragment part of the URL should be retained.
+
+    Returns:
+        A string containing the normalized URL. If normalization fails, the original URL is returned. It logs
+            a warning message in this case.
+    """
+    try:
+        # Parse the URL
+        parsed_url = urlparse(url.strip())
+        search_params = dict(parse_qsl(parsed_url.query))  # Convert query to a dict
+
+        # Remove any 'utm_' parameters
```

Comment:
Shouldn't this be more generic, `utm_` isn't the only tracking parameter I'd say?

Reply:
Basically, it's the same as in the https://github.com/apify/apify-sdk-python/pull/193#discussion_r1521200953.

@B4nan Should we add the removal of more tracking parameters (if there are any, TBH I don't know)? Or rather keep the parity.

Reply:
I'm convinced that this is way out of scope of the PR.

Reply:
Same as with the other problem, I'd say we want consistency now, if we should change something, it should be changed in the JS version too (which we surely can do as well).

@jirimoravcik do you have some suggestions on what else to ignore? I don't think we can find a "more generic solution", but we can blacklist more common parameters (at the same time we need to be careful to not blacklist something that might be used for other purposes).

Reply:
Yeah, I just wanted to point it out for discussion. E.g. for tracking, there's https://easylist.to/easylist/easyprivacy.txt